### ISSUE TITLE
Making 'href' required on links

### ISSUE THREAD

**Issue by wendellpiez** [usnistgov/OSCAL#235]
# User Story:

Whether unlinked links, in general, are a feature or a bug is debatable, but in OSCAL they are a bug:  OSCAL data does not represent data that is "wild" or uncontrolled to the point where they should be tolerable. **link** elements in our data should always have 'href' to simplify downstream processing (even if broken links are always a risk).

## Goals:

Make a 'href' a required attribute on **link** elements in the different schemas.

## Dependencies:

Published data sets containing unlinked link elements will be invalid, and need to be updated. Currently this means SP800-53 rev 4 but the issues may reappear in rev 5 as well.

Of the 26 unlinked **link** elements in the catalog presently, 23 are internal cross-references and could presumably be resolved. (If not, they represent a deeper problem.) Three (3) point to "Appendix J" (twice) and "Chapter 3". If link targets (URIs) can be identified for these, 'href' attributes can be given to them as well.

## Acceptance Criteria

1. 'link' elements are required to have 'href' attributes in OSCAL schemas (catalog, profile and metadata/header).
1. All released (sample) data sets that are intended to be valid, can be shown to be valid to the new models. This can be tested with XPath: `//link[empty(@href)]` should return an empty sequence.
1. Production pipelines for implicated data sets such as the OSCAL SP-800-53 control catalog are updated so that the missing links will be provided in future as well (by amending the data set where necessary).

**Comment by david-waltermire-nist** [comment]
@wendellpiez Has this been completed?

**Comment by wendellpiez** [comment]
No it has not. One question is determining the correct targets for all the links presently without them.

Once those have been determined (there are not many), the work entails both the production pipeline for SP800-53 and the data set(s), as well as the schemas. A lower-impact approach would be to leave link/@href optional for now and address the issue in schema documentation. (This could also make a good Schematron demo.)

**Comment by wendellpiez** [comment]
Note #346, which tracks content revisions related to this item. As noted above, those improvements to the data can be made while the validation discussed here is also managed by Schematron not by the schema. (Which cannot validate that an href value is plausible, only that the attribute is present. Unfortunately, the empty string "" is a valid URL.)

**Comment by david-waltermire-nist** [comment]
We should fix this for [link](https://pages.nist.gov/OSCAL/docs/schemas/oscal-catalog-xml/#link). For hlink, href is already required, and citations need to support the non-href case.

**Comment by iMichaela** [comment]
## 5/23/2019
@wendellpiez is working on it. Estimates to be done in a few days.

**Comment by wendellpiez** [comment]
In PR #381, the schemas have been changed to require `@href`. The SP800-53 rev4 catalog now contains a link to appendix J as follows (line 32725):

```
<link rel="related" href="https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-53r4.pdf">Appendix J</link>
```

As you can see the link goes only to a PDF copy (of the same catalog) with no reference to Appendix J. If a better link is available we should use that. @david-waltermire-nist or @iMichaela do you have any suggestions?

**Comment by wendellpiez** [comment]
After committing PR #381 we found more issues, corrected in PR #383.

**Comment by david-waltermire-nist** [comment]
Finally done.
